### ISSUE TITLE
Type var does not match on generic functions defined within classes

### ISSUE THREAD

**Issue by danmur97** [microsoft/pyright#10004]
**Description**

Generic functions defined as class properties cannot pass the pyright check, but their equivalent non-class functions do.

There are two issues:
1. pyright denies defining generic functions as properties of a class. Type variables do have meaning in properties as well as the `def` statements do.
2. calling a function with the same signature should return the same result, it does matter how it is defined (by a def statement, a property of a class, etc.)

**Code example**
```
from dataclasses import dataclass
from typing import TypeVar
from collections.abc import Callable

_T = TypeVar("_T")

@dataclass(frozen=True)
class Interface:
    function: Callable[[Callable[[bool], _T]], _T] # Pyright complains: Type variable "_T" has no meaning in this context

def function(f: Callable[[bool], _T]) -> _T:
    return f(True)

def _test_1(f: Callable[[bool], _T]) -> _T:
    return function(f) # ok!

def _test_2(i: Interface, f: Callable[[bool], _T]) -> _T:
    return i.function(f) # Pyright complains: Type "_T" is not assignable to return type "_T@_test_2"
```
Here, `function` and `i.function` are two generic functions with the same signature.
**Version**

Using pyright 1.1.392 through the CLI interface `pyright ./the_file.py`

**Comment by erictraut** [comment]
This behavior is correct and is mandated by the Python typing spec.

A type variable must have an associated scope. The scoping rules for type variables are spelled out [in this section of the spec](https://typing.readthedocs.io/en/latest/spec/generics.html#scoping-rules-for-type-variables).

Type variable scoping has traditionally been confusing in Pyright, which prompted us define and implement a cleaner syntax for Python 3.12. See [PEP 695](https://peps.python.org/pep-0695/) for details. If you are using Python 3.12 or newer, I recommend that you take advantage of the new syntax. It eliminates much of the additional boilerplate traditionally required for generics and eliminates confusion about scoping rules.

I suspect that your intent in the above code was to scope `_T` to the `Interface` class. You can do this using the traditional syntax:

```python
@dataclass(frozen=True)
class Interface(Generic[_T]):
    function: Callable[[Callable[[bool], _T]], _T]
```

Or using the new 3.12 syntax:

```python
@dataclass(frozen=True)
class Interface[T]:
    function: Callable[[Callable[[bool], T]], T]
```

**Comment by danmur97** [comment]
@erictraut That workaround of making the class generic is what I intend to avoid. Also, by the same sections of the spec it says it's not illegal to use unscooped type vars on methods.

![Image](https://github.com/user-attachments/assets/e7116334-211e-4e3d-bb7d-6e2a7e8ce907)

**Comment by erictraut** [comment]
You are not defining a method in your code sample. You're defining an instance variable that is annotated with a `Callable` type form. A type variable must be scoped to a function declaration, a class declaration, or a type alias declaration. An unscoped type variable can't appear in random type expressions like `list[T]` or `Callable[[bool], T]`.

**Comment by erictraut** [comment]
If your intent is to define a generic static method `Interface.method`, you can do so using this approach:

```python
@dataclass(frozen=True)
class Interface:
    @staticmethod
    def function[T](x: Callable[[bool], T]) -> T: ...
```

**Comment by danmur97** [comment]
That's not longer an interface, using `def` forces to define the method implementation.
Class instances are _unspecified_ values of some type.
e.g.
```
@dataclass(frozen=True)
class Point:
    x: int
    y: int
```
I can instantiate `Point` by specifying its concrete values `p = Point(2, 3)` 
```
@dataclass(frozen=True)
class Interface:
    x: Callable[[int], int]
```
Callable are no special types, they are also types as well. They can be also be instantiated. Using lambda (or def) `i = Interface(lambda x: x + 1)`

```
generic: Callable[[bool, _T], _T] = lambda b, t: t

def generic2(b: bool, t: _T) -> _T:
    return t
reveal_type(generic) # Pyright says: Type of "generic" is "(b: bool, t: _T) -> _T"
reveal_type(generic2)  # Pyright says:  Type of "generic2" is "(b: bool, t: _T@generic2) -> _T@generic2"Pyright
```
`generic` and `generic2` are two equivalent ways to declare the same function. If _T were a static type like `int` there is no problem. Why the generic case should be an exception? Why lock to the `def` statement when lambda is a perfectly valid and equivalent style of instantiation?

**Comment by danmur97** [comment]
I know that abstract methods/classes exist and that they will be the workaround to fulfil the interface definition without this issue. However, I am using a functional approach and avoiding any type of inheritance (that an abstract class will force).
Having functions as instance variables avoids abstract and object-oriented stuff.

**Comment by danmur97** [comment]
Other thing I would like to remark is that this does not happen on other type checkers (i.e. mypy). They allow this generic style, so even if it is intended to not-fixing, I would appreciate some kind of config or ignore flag where this behavior can be selected (I don't know if it already exists).

**Comment by erictraut** [comment]
There were many discussions about this very topic when we were designing PEP 695. I personally led many of those discussions and was the primary author of that PEP, so I consider myself an authority in this area.

As I said above, pyright conforms with the Python typing spec here. If you would like to suggest changes to the typing spec, the process is documented [here](https://github.com/python/typing-council#decisions).

If your intent is to define an interface, a structural type (a protocol or an abstract class) would be more appropriate. What you're defining above is a nominal class with an instance variable, and a type variable cannot be scoped to an instance variable.

> Other thing I would like to remark is that this does not happen on other type checkers (i.e. mypy)

Yes, mypy does not conform with the typing spec in this regard. There are [numerous areas](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html) where it does not fully comply with the spec. 

We've been making efforts over the past two years to clarify behaviors that were previously ambiguous, and the mypy maintainers have been making slow-but-steady progress in updating mypy accordingly, but they are reluctant to make changes that break existing typed code bases.

**Comment by erictraut** [comment]
It occurs to me that you can accomplish your goal by defining a callback protocol.

```python
from dataclasses import dataclass
from typing import Protocol, Callable

class Func(Protocol):
    def __call__[T](self, x: Callable[[bool], T], /) -> T: ...

@dataclass(frozen=True)
class Interface:
    function: Func

def test[T](i: Interface, f: Callable[[bool], T]) -> T:
    return i.function(f)
```
